doc/reference/usermode/memory_domain.rst

```diff
@@ -1,38 +1,288 @@
-but are also useful for debugging (unexpected access would cause an exception).
-
-Since architectures generally have constraints on how many partitions can be
-defined, and the size/alignment of each partition, users may need to group
-related data together using linker sections.
-
-.. contents::
-    :local:
-    :depth: 2
-
-Concepts
-********
-
-A memory domain contains some number of memory partitions.
-A memory partition is a memory region (might be RAM, peripheral registers,
-or flash, for example) with specific attributes (access permission, e.g.
-privileged read/write, unprivileged read-only, or execute never).
-Memory partitions are defined by a set of underlying MPU regions
-or MMU tables. A thread belongs to a single memory domain at
-any point in time but a memory domain may contain multiple threads.
-Threads in the same memory domain have the same access permissions
-to the memory partitions belonging to the memory domain. New threads
-will inherit any memory domain configuration from the parent thread.
-
-Implementation
+Overview
+########
+
+Zephyr's memory protection design is geared towards microcontrollers with MPU
+(Memory Protection Unit) hardware. We do support some architectures which have
+a paged MMU (Memory Management Unit), but in that case the MMU is used like
+an MPU with an identity page table.
+
+All of the discussion below will be using MPU terminology; systems with MMUs
+can be considered to have an MPU with an unlimited number of programmable
+regions.
+
+There are a few different levels on how memory access is configured when
+Zephyr memory protection features are enabled, which we will describe here:
+
+Boot Time Memory Configuration
+******************************
+
+This is the configuration of the MPU when the kernel has started up. It should
+contain the following:
+
+- Any configuration of memory regions which need to have special caching or
+  write-back policies for basic hardware and driver function. Note that most
+  MPUs have the concept of a default memory access policy map, which can be
+  enabled as a "background" mapping for any area of memory that doesn't
+  have an MPU region configuring it. It is strongly recommended to use this
+  to maximize the number of available MPU regions for the end user. On
+  ARMv7-M this is called the System Address Map.
+
+- A read-only, executable region or regions for program text and ro-data, that
+  is accessible to user mode. This could be further sub-divided into a
+  read-only region for rodata, and a read-only, executable region for text, but
+  this will require an additional MPU region. This is required so that
+  threads running in user mode can read rodata and fetch instructions.
+
+Assuming a background map which allows supervisor mode to access any memory
+it needs, and regions which grant user mode access to text/rodata, this is
+sufficient for the boot time configuration.
+
```

Comment:
Certain configurations have also PRV-RW/nPRIV-RW fixed (static, i.e. programmed at boot) regions for GCOV, and HEP as well - do we need to mention here?